README.md
# hub.docker.com/tiredofit/sentry

# Introduction

Dockerfile to build a [Sentry](https://www.sentry.io) container image.

* This Container uses Debian Stretch as a base which includes [s6 overlay](https://github.com/just-containers/s6-overlay) enabled for PID 1 Init capabilities, [zabbix-agent](https://zabbix.org) based on 3.4 Packages for individual container monitoring, Cron also installed along with other tools (bash,curl, less, logrotate, mariadb-client, nano, vim) for easier management.

* LDAP Functionality included
* SSO Functionality included

[Changelog](CHANGELOG.md)

# Authors

- [Dave Conroy](http://github/tiredofit/)

# Table of Contents

- [Introduction](#introduction)
    - [Changelog](CHANGELOG.md)
- [Prerequisites](#prerequisites)
- [Installation](#installation)
- [Quick Start](#quick-start)
- [Configuration](#configuration)
    - [Database](#database)
    - [Data Volumes](#data-volumes)
    - [Environment Variables](#environmentvariables)   
    - [Networking](#networking)
- [Maintenance](#maintenance)
    - [Shell Access](#shell-access)
   - [References](#references)

# Prerequisites

This image assumes that you are using a reverse proxy such as [jwilder/nginx-proxy](https://github.com/jwilder/nginx-proxy) and optionally the [Let's Encrypt Proxy Companion @ https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion) in order to serve your pages. However, it will run just fine on it's own if you map appropriate ports.

You will need a Redis server and also a Postgresql 9.6 DB Server


# Installation

Automated builds of the image are available on [Registry](https://hub.docker.com/r/tiredofit/sentry) and is the recommended method of installation.


```bash
docker pull /tiredofit/sentry:(imagetag)
```

The following image tags are available:

* `latest` - Most recent release of Sentry w/most recent Debian Stretch

# Quick Start

* The quickest way to get started is using [docker-compose](https://docs.docker.com/compose/). See the examples folder for a working [docker-compose.yml](examples/docker-compose.yml) that can be modified for development or production use.

* Set various [environment variables](#environment-variables) to understand the capabilities of this image.
* Map [persistent storage](#data-volumes) for access to configuration and data files for backup.

# Configuration

### Data-Volumes

The following directories are used for configuration and can be mapped for persistent storage.

| Directory | Description |
|-----------|-------------|
| `/etc/sentry` | If you would like to use your own configuration, expose this folder. |
| `/var/lib/sentry/files` | Sentry Data |

### Database

An external Postgresql Database is required for using this image.
Additionally, you must have access to a Redis Server

### Environment Variables

Along with the Environment Variables from the Base image (tiredfoit/debian:stretch), below is the complete list of available options 
that can be used to customize your installation.

| Parameter | Description |
|-----------|-------------|
| `DB_HOST` | Name of Postgresql Container e.g. `sentry-db`|
| `DB_PORT` | Port of Postgresql Container Default `5432`|
| `DB_NAME` | Name of Postgresql Database |
| `DB_USER` | Username for Postgresql Database |
| `DB_PASS` | Username for Postgresql Database |
| `RABBITMQ_HOST` | RabbitMQ Hostname if utilized |
| `RABBITMQ_USER` | Username for Postgresql Database|
| `RABBITMQ_PASS` | Password for RabbitMQ if used |
| `RABBITMQ_VHOST` | RabbitMQ Vhostname if used |
| `REDIS_HOST` | Redis Server Container e.g. `sentry-redis` |
| `REDIS_PASS` | Redis Password if used |
| `REDIS_PORT` | Redis Port number Default `6379` |
| `REDIS_DB` | Redis DB Number if required |
| `MEMCACHED_HOST` | Memcached Container name e.g. `sentry-memcached` |
| `MEMCACHED_PORT` | Memcached Port Default `11211` |
| `SERVER_EMAIL` | Default Server FROM address |
| `SMTP_HOST` | SMTP Server Hostname |
| `SMTP_PORT` | SMTP Server Port |
| `SMTP_USER` | Username for SMTP Server if used |
| `SMTP_PASS` | Password for SMTP Server if used |
| `SMTP_USE_TLS` | Use TLS with SMTP `True` or `False` Default `False` |
| `ENABLE_EMAIL_REPLIES` | Enable Email Replies `True` or `False` Default `False` |
| `SMTP_HOSTNAME` | What is the hostname you with to advertise yourself as .. |
| `MAILGUN_API_KEY` | Mailgun API key if used |
| `SINGLE_ORGANIZATION` | Enable Single Organizations `True`/`False` Default `True` |
| `SECRET_KEY` | 32 character long secret key |
| `GITHUB_APP_ID` | If using Github, enter AppID |
| `GITHUB_API_SECRET` | API for Github |
| `BITBUCKET_CONSUMER_KEY` | If using Bitbucket, enter Key |
| `BITBUCKET_CONSUMER_SECRET` | Bitbucket Key |

LDAP / SSO Functionality:


| Parameter | Description |
|-----------|-------------|
| `LDAP_ENABLE` | Enable Ldap Support `TRUE`/`FALSE` Default `FALSE` |
| `LDAP_BASE_DN` | LDAP Base DN e.g. `dc=example,dc=com` |
| `LDAP_BIND_USER` | Bind user e.g. `cn=admin,dc=example,dc=com` | 
| `LDAP_BIND_PASSWORD` | LDAP Bind Password |
| `LDAP_URI` | URI of LDAP Server e.g. `ldap://ldap.example.com` |
| `LDAP_DOMAIN` | @suffix for members who don't have `mail` attribute e.g. `example.com` |
| `LDAP_FIND_GROUP_PERMS` | Find Group Permisions `True`/`False` Default `True` |
| `LDAP_CACHE_GROUPS` | Cache Groups `True`/`False` Default `True`
| `LDAP_GROUP_CACHE_TIMEOUT` | Group Cache Timeout in seconds Default: `3600` |
| `LDAP_UID_ATTR` | UID attribute Default `uid` |
| `LDAP_NAME_ATTR` | Name Attribute Default `cn` |
| `LDAP_MAIL_ATTR` | Mail Attribute Default `mail` |
| `LDAP_USER_SCOPE` | User Search scope Default `(mail=%(user)s)` |
| `LDAP_GROUP_SCOPE` |  Group Search Scope Default `(objectClass=groupOfUniqueNames)` |
| `LDAP_ORGANIZATION_NAME` | Organization Name to place members in Default `sentry` |
| `LDAP_AUTO_ADD_ROLE_TYPE` |  Role type auto added users are assigned Default `member`
| `LDAP_AUTO_ADD_GLOBAL` |  Autocreated users should be granted global access to org Default `False`


### Networking

The following ports are exposed.

| Port      | Description |
|-----------|-------------|
| `9000`    | Sentry UWSGI|

# Maintenance
#### Shell Access

For debugging and maintenance purposes you may want access the containers shell. 

```bash
docker exec -it (whatever your container name is e.g. sentry) bash
```

# References

* https://www.sentry.io




CHANGELOG.md
## 1.2 2018-04-25 <dave at tiredofit dot ca>

* Add Cron Support for Cleanup

## 1.1 2018-04-25 <dave at tiredofit dot ca>

* Add LDAP Support
* Add SSO/SAML Support
* Cleanup Image Size

## 1.0 2018-04-23 <dave at tiredofit dot ca>

* Initial Release
* Debian Stretch
* Sentry 8.22


install/assets/etc/sentry/sentry.conf.py
# This file is just Python, with a touch of Django which means
# you can inherit and tweak settings to your hearts content.

# For Docker, the following environment variables are supported:
#  DB_HOST
#  DB_PORT
#  DB_NAME
#  DB_USER
#  DB_PASS
#  RABBITMQ_HOST
#  RABBITMQ_USER
#  RABBITMQ_PASS
#  RABBITMQ_VHOST
#  REDIS_HOST
#  REDIS_PASSW
#  REDIS_PORT
#  REDIS_DB
#  MEMCACHED_HOST
#  MEMCACHED_PORT
#  SENTRY_FILESTORE_DIR
#  SERVER_EMAIL
#  SMTP_HOST
#  SMTP_PORT
#  SMTP_USER
#  SMTP_PASS
#  SMTP_USE_TLS
#  ENABLE_EMAIL_REPLIES
#  SMTP_HOSTNAME
#  MAILGUN_API_KEY
#  SINGLE_ORGANIZATION
#  SECRET_KEY
#  GITHUB_APP_ID
#  GITHUB_API_SECRET
#  BITBUCKET_CONSUMER_KEY
#  BITBUCKET_CONSUMER_SECRET
from sentry.conf.server import *  # NOQA

import os
import os.path

CONF_ROOT = os.path.dirname(__file__)

postgres = env('DB_HOST') or (env('DB_PORT') and 'postgres')
if postgres:
    DATABASES = {
        'default': {
            'ENGINE': 'sentry.db.postgres',
            'NAME': (
                env('SENTRY_DB_NAME')
                or env('DB_NAME')
                or 'postgres'
            ),
            'USER': (
                env('SENTRY_DB_USER')
                or env('DB_USER')
                or 'postgres'
            ),
            'PASSWORD': (
                env('SENTRY_DB_PASSWORD')
                or env('DB_PASS')
                or ''
            ),
            'HOST': postgres,
            'PORT': (
                env('DB_PORT')
                or ''
            ),
            'OPTIONS': {
                'autocommit': True,
            },
        },
    }

# You should not change this setting after your database has been created
# unless you have altered all schemas first
SENTRY_USE_BIG_INTS = True

# If you're expecting any kind of real traffic on Sentry, we highly recommend
# configuring the CACHES and Redis settings

###########
# General #
###########

# Instruct Sentry that this install intends to be run by a single organization
# and thus various UI optimizations should be enabled.
SENTRY_SINGLE_ORGANIZATION = env('SINGLE_ORGANIZATION', True)

#########
# Redis #
#########

# Generic Redis configuration used as defaults for various things including:
# Buffers, Quotas, TSDB

redis = env('REDIS_HOST') or (env('REDIS_PORT') and 'redis')
if not redis:
    raise Exception('Error: REDIS_PORT (or REDIS_HOST) is undefined, did you forget to `--link` a redis container?')

redis_password = env('REDIS_PASS') or ''
redis_port = env('REDIS_PORT') or '6379'
redis_db = env('REDIS_DB') or '0'

SENTRY_OPTIONS.update({
    'redis.clusters': {
        'default': {
            'hosts': {
                0: {
                    'host': redis,
                    'password': redis_password,
                    'port': redis_port,
                    'db': redis_db,
                },
            },
        },
    },
})

#########
# Cache #
#########

# Sentry currently utilizes two separate mechanisms. While CACHES is not a
# requirement, it will optimize several high throughput patterns.

memcached = env('MEMCACHED_HOST') or (env('MEMCAHCED_PORT') and 'memcached')
if memcached:
    memcached_port = (
        env('MEMCACHED_PORT')
        or '11211'
    )
    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
            'LOCATION': [memcached + ':' + memcached_port],
            'TIMEOUT': 3600,
        }
    }

# A primary cache is required for things such as processing events
SENTRY_CACHE = 'sentry.cache.redis.RedisCache'

#########
# Queue #
#########

# See https://docs.getsentry.com/on-premise/server/queue/ for more
# information on configuring your queue broker and workers. Sentry relies
# on a Python framework called Celery to manage queues.

rabbitmq = env('RABBITMQ_HOST') or (env('RABBITMQ_PORT') and 'rabbitmq')

if rabbitmq:
    BROKER_URL = (
        'amqp://' + (
            env('RABBITMQ_USER')
            or env('RABBITMQ_ENV_RABBITMQ_DEFAULT_USER')
            or 'guest'
        ) + ':' + (
            env('RABBITMQ_PASS')
            or env('RABBITMQ_ENV_RABBITMQ_DEFAULT_PASS')
            or 'guest'
        ) + '@' + rabbitmq + '/' + (
            env('RABBITMQ_VHOST')
            or env('RABBITMQ_ENV_RABBITMQ_DEFAULT_VHOST')
            or '/'
        )
    )
else:
    BROKER_URL = 'redis://:' + redis_password + '@' + redis + ':' + redis_port + '/' + redis_db


###############
# Rate Limits #
###############

# Rate limits apply to notification handlers and are enforced per-project
# automatically.

SENTRY_RATELIMITER = 'sentry.ratelimits.redis.RedisRateLimiter'

##################
# Update Buffers #
##################

# Buffers (combined with queueing) act as an intermediate layer between the
# database and the storage API. They will greatly improve efficiency on large
# numbers of the same events being sent to the API in a short amount of time.
# (read: if you send any kind of real data to Sentry, you should enable buffers)

SENTRY_BUFFER = 'sentry.buffer.redis.RedisBuffer'

##########
# Quotas #
##########

# Quotas allow you to rate limit individual projects or the Sentry install as
# a whole.

SENTRY_QUOTAS = 'sentry.quotas.redis.RedisQuota'

########
# TSDB #
########

# The TSDB is used for building charts as well as making things like per-rate
# alerts possible.

SENTRY_TSDB = 'sentry.tsdb.redis.RedisTSDB'

###########
# Digests #
###########

# The digest backend powers notification summaries.

SENTRY_DIGESTS = 'sentry.digests.backends.redis.RedisBackend'

################
# File storage #
################

# Uploaded media uses these `filestore` settings. The available
# backends are either `filesystem` or `s3`.

SENTRY_OPTIONS['filestore.backend'] = 'filesystem'
SENTRY_OPTIONS['filestore.options'] = {
    'location': env('SENTRY_FILESTORE_DIR'),
}

##############
# Web Server #
##############

# If you're using a reverse SSL proxy, you should enable the X-Forwarded-Proto
# header and set `SENTRY_USE_SSL=1`

if env('SENTRY_USE_SSL', False):
    SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTO', 'https')
    SESSION_COOKIE_SECURE = True
    CSRF_COOKIE_SECURE = True
    SOCIAL_AUTH_REDIRECT_IS_HTTPS = True

SENTRY_WEB_HOST = '0.0.0.0'
SENTRY_WEB_PORT = 9000
SENTRY_WEB_OPTIONS = {
    'workers': int(env('SENTRY_WORKERS') or 3),
    'limit_request_line': 0,  # required for raven-js
    'errorlog' : '/var/log/sentry/sentry_error.log',
    'accesslog' : '/var/log/sentry/sentry_access.log',
}

###############
# Mail Server #
###############
email = env('SMTP_HOST') or (env('SMTP_PORT') and 'smtp')
if email:
    SENTRY_OPTIONS['mail.backend'] = 'smtp'
    SENTRY_OPTIONS['mail.host'] = email
    SENTRY_OPTIONS['mail.password'] = env('SMTP_PASS') or ''
    SENTRY_OPTIONS['mail.username'] = env('SMTP_USER') or ''
    SENTRY_OPTIONS['mail.port'] = int(env('SMTP_PORT') or 25)
    SENTRY_OPTIONS['mail.use-tls'] = env('SMTP_USE_TLS', False)
else:
    SENTRY_OPTIONS['mail.backend'] = 'dummy'

# The email address to send on behalf of
SENTRY_OPTIONS['mail.from'] = env('SMTP_MAIL_FROM') or 'root@localhost'

# If you're using mailgun for inbound mail, set your API key and configure a
# route to forward to /api/hooks/mailgun/inbound/
SENTRY_OPTIONS['mail.mailgun-api-key'] = env('MAILGUN_API_KEY') or ''

# If you specify a MAILGUN_API_KEY, you definitely want EMAIL_REPLIES
if SENTRY_OPTIONS['mail.mailgun-api-key']:
    SENTRY_OPTIONS['mail.enable-replies'] = True
else:
    SENTRY_OPTIONS['mail.enable-replies'] = env('ENABLE_EMAIL_REPLIES', False)

if SENTRY_OPTIONS['mail.enable-replies']:
    SENTRY_OPTIONS['mail.reply-hostname'] = env('SMTP_HOSTNAME') or ''

# If this value ever becomes compromised, it's important to regenerate your
# SENTRY_SECRET_KEY. Changing this value will result in all current sessions
# being invalidated.
secret_key = env('SECRET_KEY')
if not secret_key:
    raise Exception('Error: SECRET_KEY is undefined, run `generate-secret-key` and set to -e SECRET_KEY')

if 'SENTRY_RUNNING_UWSGI' not in os.environ and len(secret_key) < 32:
    print('!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!')
    print('!!                    CAUTION                       !!')
    print('!! Your SECRET_KEY is potentially insecure.         !!')
    print('!!    We recommend at least 32 characters long.     !!')
    print('!!     Regenerate with `generate-secret-key`.       !!')
    print('!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!')

SENTRY_OPTIONS['system.secret-key'] = secret_key
LOGGING['disable_existing_loggers'] = False
SENTRY_BEACON = env('BEACON', False)
SENTRY_FEATURES['auth:register'] = env('ALLOW_REGISTRATION',True)


if 'GITHUB_APP_ID' in os.environ:
    GITHUB_EXTENDED_PERMISSIONS = ['repo']
    GITHUB_APP_ID = env('GITHUB_APP_ID')
    GITHUB_API_SECRET = env('GITHUB_API_SECRET')

if 'BITBUCKET_CONSUMER_KEY' in os.environ:
    BITBUCKET_CONSUMER_KEY = env('BITBUCKET_CONSUMER_KEY')
    BITBUCKET_CONSUMER_SECRET = env('BITBUCKET_CONSUMER_SECRET')

SENTRY_FEATURES['organizations:sso'] = False
SENTRY_FEATURES['organizations:sso-saml2'] = False
SENTRY_FEATURES['organizations:sso-rippling'] = False
